pkg/services/ngalert/api/prometheus/api_prometheus.go

```diff
@@ -429,6 +430,177 @@ func RuleAlertStateMutatorGenerator(manager state.AlertInstanceManager) RuleAler
+			SearchRuleGroup: ctx.searchRuleGroup,
+		},
+		RuleType:      ctx.ruleType,
+		Limit:         remainingGroups,
+		RuleLimit:     remainingRules,
+		ContinueToken: token,
+	}
+
+	ruleList, newToken, err := store.ListAlertRulesByGroup(ctx.opts.Ctx, &byGroupQuery)
+	if err != nil {
+		return pageResult{}, err
+	}
+
+	span.SetAttributes(
+		attribute.Int("store_rule_list_len", len(ruleList)),
+		attribute.Bool("store_continue_token_set", newToken != ""),
+	)
+	span.AddEvent("Alert rules retrieved from store")
+
+	groupedRules := getGroupedRules(log, ruleList, ctx.ruleNamesSet, ctx.opts.AllowedNamespaces)
+
+	result := pageResult{
+		groups:      make([]apimodels.RuleGroup, 0, len(groupedRules)),
+		totalsDelta: make(map[string]int64),
+		nextToken:   newToken,
+		hasMore:     newToken != "",
+	}
+
+	for _, rg := range groupedRules {
+		ruleGroup, totals := toRuleGroup(
+			log, rg.GroupKey, rg.Folder, rg.Rules,
+			ctx.provenanceRecords, ctx.limitAlertsPerRule,
+			ctx.stateFilterSet, ctx.matchers, ctx.labelOptions,
+			ctx.ruleStatusMutator, ctx.alertStateMutator,
+		)
+		ruleGroup.Totals = totals
+		accumulateTotals(result.totalsDelta, totals)
+
+		if len(ctx.stateFilterSet) > 0 {
+			filterRulesByState(ruleGroup, ctx.stateFilterSet)
+		}
+
+		if len(ctx.healthFilterSet) > 0 {
+			filterRulesByHealth(ruleGroup, ctx.healthFilterSet)
+		}
+
+		if ctx.limitRulesPerGroup > -1 && int64(len(ruleGroup.Rules)) > ctx.limitRulesPerGroup {
+			ruleGroup.Rules = ruleGroup.Rules[0:ctx.limitRulesPerGroup]
+		}
+
+		if len(ruleGroup.Rules) > 0 {
+			result.groups = append(result.groups, *ruleGroup)
+		}
+	}
+
+	return result, nil
+}
+
+// paginateRuleGroups fetches pages until limits are satisfied applying filters at each step
+func paginateRuleGroups(log log.Logger, store ListAlertRulesStoreV2, ctx *paginationContext, span trace.Span, maxGroups, maxRules int64, startToken string) ([]apimodels.RuleGroup, map[string]int64, string, error) {
+	allGroups := []apimodels.RuleGroup{}
+	rulesTotals := make(map[string]int64)
+
+	continueToken := startToken
+	groupsReturned := int64(0)
+	rulesReturned := int64(0)
+
+	for {
+		remainingGroups := maxGroups
+		if maxGroups > 0 {
+			remainingGroups = maxGroups - groupsReturned
+			if remainingGroups <= 0 {
+				break
+			}
+		}
+		remainingRules := maxRules
+		if maxRules > 0 {
+			remainingRules = maxRules - rulesReturned
+			if remainingRules <= 0 {
+				break
+			}
+		}
+
+		page, err := ctx.fetchAndFilterPage(log, store, span, continueToken, remainingGroups, remainingRules)
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		accumulateTotals(rulesTotals, page.totalsDelta)
+
+		// Add groups and check limits
+		for _, group := range page.groups {
+			allGroups = append(allGroups, group)
+			groupsReturned++
+			rulesReturned += int64(len(group.Rules))
+
+			// Check if we've hit limits
+			if (maxGroups > 0 && groupsReturned >= maxGroups) || (maxRules > 0 && rulesReturned >= maxRules) {
```

Comment:
> `groupsReturned >= maxGroups`

Shall we make this strict equality?

Reply:
Updated: https://github.com/grafana/grafana/pull/114334/commits/3529b19ae0244e6d572771c8af2dd8f12635ebe4